leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/CaliforniumLwM2mRequestSender.java

```diff
@@ -137,34 +139,41 @@ public T buildResponse(final Response coapResponse) {
         endpoint.sendRequest(coapRequest);
     }
 
-    public synchronized void cancelPendingRequests(String registrationId) {
-        ArrayList<Request> requests = pendingRequests.get(registrationId);
-        if (requests != null) {
-            for (Request coapRequest : requests) {
-                coapRequest.cancel();
-                removePendingRequest(registrationId, coapRequest);
-            }
-        }
+    private String getFloorKey(String registrationId) {
+        if (registrationId == null)
+            return null;
+        return registrationId + '#';
     }
 
-    private synchronized void addPendingRequest(String registrationId, Request coapRequest) {
-        // Get or create all pending requests for this registrationId
-        ArrayList<Request> requests = pendingRequests.get(registrationId);
-        if (requests == null) {
-            requests = new ArrayList<>();
-            pendingRequests.put(registrationId, requests);
+    private String getCeilingKey(String registrationId) {
+        if (registrationId == null)
+            return null;
+        return registrationId + "#A";
+    }
+
+    private String getKey(String registrationId, Request coapRequest) {
+        if (registrationId == null || coapRequest == null)
+            return null;
+        return registrationId + '#' + coapRequest.hashCode();
```

Comment:
Though `addPendingRequest` always creates a `CleanerMessageObserver`, it should not be too hard to move the "key" (reg-id + unique id) to that class and then use that key in add and remove.
So there is no need to depend on `hashCode` of a third party library (even if we know, it's not really third party :-).